Code folding is buggy

Sometimes folding ranges are out of sync and located 1 line below or upwards. Sometimes they are completely broken with following exception:
```
[Error - 18:29:56] 27 апр. 2020 г., 18:29:56 Problem with folding range for XXX.java:
Invalid_Char_In_String
org.eclipse.jdt.core.compiler.InvalidInputException: Invalid_Char_In_String
	at org.eclipse.jdt.internal.core.util.PublicScanner.getNextToken(PublicScanner.java:1564)
	at org.eclipse.jdt.ls.core.internal.handlers.FoldingRangeHandler.computeFoldingRanges(FoldingRangeHandler.java:112)
	at org.eclipse.jdt.ls.core.internal.handlers.FoldingRangeHandler.foldingRange(FoldingRangeHandler.java:63)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$27(JDTLanguageServer.java:814)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$49(JDTLanguageServer.java:950)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

STR:
1. Add some parameter to the method used across class
2. For some time method call will be causing compile error
3. Fix it by adding desired argument
4. Code folding is broken!!!

Yes, it's vscode extension's version.

More traces (not sure if they are related) Even without splitting editor foldings disappear after some edits:
```

[Error - 19:14:23] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:645)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:373)
	at java.base/java.util.ArrayList.get(ArrayList.java:425)
	at org.eclipse.jdt.core.dom.ASTNode$NodeList.get(ASTNode.java:1472)
	at org.eclipse.jdt.internal.corext.refactoring.code.SourceProvider.getParameterData(SourceProvider.java:316)
	at org.eclipse.jdt.internal.corext.refactoring.code.CallInliner.computeRealArguments(CallInliner.java:474)
	at org.eclipse.jdt.internal.corext.refactoring.code.CallInliner.initialize(CallInliner.java:256)
	at org.eclipse.jdt.internal.corext.refactoring.code.InlineMethodRefactoring.checkFinalConditions(InlineMethodRefactoring.java:311)
	at org.eclipse.ltk.core.refactoring.CheckConditionsOperation.run(CheckConditionsOperation.java:86)
	at org.eclipse.ltk.core.refactoring.CreateChangeOperation.run(CreateChangeOperation.java:122)
	at org.eclipse.jdt.ls.core.internal.corrections.RefactorProcessor.getInlineProposal(RefactorProcessor.java:353)
	at org.eclipse.jdt.ls.core.internal.corrections.RefactorProcessor.getProposals(RefactorProcessor.java:143)
	at org.eclipse.jdt.ls.core.internal.handlers.CodeActionHandler.getCodeActionCommands(CodeActionHandler.java:121)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$14(JDTLanguageServer.java:620)
	at org.eclipse.jdt.ls.core.internal.BaseJDTLanguageServer.lambda$0(BaseJDTLanguageServer.java:74)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	... 6 more

[Error - 19:15:02] 27 апр. 2020 г., 19:15:02 Problems occurred when invoking code from plug-in: "org.eclipse.core.resources".
null
java.lang.NullPointerException
```

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FoldingRangeHandler.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FoldingRangeHandler.java
@@ -82,7 +82,7 @@ public class FoldingRangeHandler {
 			scanner.resetTo(shift, shift + range.getLength());
 
 			int start = shift;
-			int token = scanner.getNextToken();
+			int token = 0;
 			Stack<Integer> regionStarts = new Stack<>();
 			while (token != ITerminalSymbols.TokenNameEOF) {
 				start = scanner.getCurrentTokenStartPosition();
@@ -109,19 +109,29 @@ public class FoldingRangeHandler {
 					default:
 						break;
 				}
-				token = scanner.getNextToken();
+				token = getNextToken(scanner);
 			}
-
 			computeTypeRootRanges(foldingRanges, unit, scanner);
-		} catch (CoreException |
-
-				InvalidInputException e) {
+		} catch (CoreException e) {
 			JavaLanguageServerPlugin.logException("Problem with folding range for " + unit.getPath().toPortableString(), e);
 			monitor.setCanceled(true);
 		}
 	}
 
-	private void computeTypeRootRanges(List<FoldingRange> foldingRanges, ITypeRoot unit, IScanner scanner) throws CoreException, InvalidInputException {
+	private int getNextToken(IScanner scanner) {
+		int token = 0;
+		while (token == 0) {
+			try {
+				token = scanner.getNextToken();
+			} catch (InvalidInputException e) {
+				// ignore
+				// JavaLanguageServerPlugin.logException("Problem with folding range", e);
+			}
+		}
+		return token;
+	}
+
+	private void computeTypeRootRanges(List<FoldingRange> foldingRanges, ITypeRoot unit, IScanner scanner) throws CoreException {
 		if (unit.hasChildren()) {
 			for (IJavaElement child : unit.getChildren()) {
 				if (child instanceof IImportContainer) {
@@ -136,7 +146,7 @@ public class FoldingRangeHandler {
 		}
 	}
 
-	private void computeTypeRanges(List<FoldingRange> foldingRanges, IType unit, IScanner scanner) throws CoreException, InvalidInputException {
+	private void computeTypeRanges(List<FoldingRange> foldingRanges, IType unit, IScanner scanner) throws CoreException {
 		ISourceRange typeRange = unit.getSourceRange();
 		foldingRanges.add(new FoldingRange(scanner.getLineNumber(unit.getNameRange().getOffset()) - 1, scanner.getLineNumber(typeRange.getOffset() + typeRange.getLength()) - 1));
 		IJavaElement[] children = unit.getChildren();
@@ -149,7 +159,7 @@ public class FoldingRangeHandler {
 		}
 	}
 
-	private void computeMethodRanges(List<FoldingRange> foldingRanges, IMethod method, IScanner scanner) throws CoreException, InvalidInputException {
+	private void computeMethodRanges(List<FoldingRange> foldingRanges, IMethod method, IScanner scanner) throws CoreException {
 		ISourceRange sourceRange = method.getSourceRange();
 		final int shift = sourceRange.getOffset();
 		scanner.resetTo(shift, shift + sourceRange.getLength());
@@ -157,7 +167,7 @@ public class FoldingRangeHandler {
 		foldingRanges.add(new FoldingRange(scanner.getLineNumber(method.getNameRange().getOffset()) - 1, scanner.getLineNumber(shift + sourceRange.getLength()) - 1));
 
 		int start = shift;
-		int token = scanner.getNextToken();
+		int token = 0;
 		Stack<Integer> leftParens = null;
 		int prevCaseLine = -1;
 		Map<Integer, Integer> candidates = new HashMap<>();
@@ -208,7 +218,7 @@ public class FoldingRangeHandler {
 				default:
 					break;
 			}
-			token = scanner.getNextToken();
+			token = getNextToken(scanner);
 		}
 
 		for (Map.Entry<Integer, Integer> entry : candidates.entrySet()) {

--- a/org.eclipse.jdt.ls.tests/projects/maven/foldingRange/src/main/java/org/sample/InvalidInputRange.java
+++ b/org.eclipse.jdt.ls.tests/projects/maven/foldingRange/src/main/java/org/sample/InvalidInputRange.java
@@ -1,0 +1,11 @@
+package org.sample;
+
+/**
+ * InvalidInputRange
+ */
+public class InvalidInputRange {
+
+    void foo() {
+        char c = '';
+    }
+}

--- a/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/FoldingRangeHandlerTest.java
+++ b/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/FoldingRangeHandlerTest.java
@@ -71,6 +71,16 @@ public class FoldingRangeHandlerTest extends AbstractProjectsManagerBasedTest {
 	}
 
 	@Test
+	public void testInvalidInput() throws Exception {
+		String className = "org.sample.InvalidInputRange";
+		List<FoldingRange> foldingRanges = getFoldingRanges(className);
+		assertTrue(foldingRanges.size() == 3);
+		assertHasFoldingRange(2, 4, "comment", foldingRanges);
+		assertHasFoldingRange(5, 10, null, foldingRanges);
+		assertHasFoldingRange(7, 9, null, foldingRanges);
+	}
+
+	@Test
 	public void testRegionFoldingRanges() throws Exception {
 		String className = "org.sample.RegionFoldingRange";
 		List<FoldingRange> foldingRanges = getFoldingRanges(className);

